File: framework/C/src/pool.c
/* pool.c -- oligo pool and primer-based random access */
#include "vivi/pool.h"
#include "vivi/genome.h"
#include <string.h>

/* absorbs messages when a caller passes err == nullptr */
static const char *pool_err_sink;

void vivi_pool_free(vivi_pool *p)
{
	if (!p) return;
	for (size_t i = 0; i < p->count; i++) vivi_dealloc(p->strands[i]);
	vivi_dealloc(p->ids);
	vivi_dealloc(p->strands);
	vivi_dealloc(p->lens);
	p->ids = nullptr;
	p->strands = nullptr;
	p->lens = nullptr;
	p->count = 0;
	p->cap = 0;
}

static bool pool_reserve(vivi_pool *p, size_t need)
{
	if (p->cap >= need) return true;
	size_t cap = p->cap ? p->cap * 2 : 8;
	if (cap < need) cap = need;
	int *ids = vivi_alloc(cap * sizeof(int));
	uint8_t **strands = vivi_alloc(cap * sizeof(uint8_t *));
	size_t *lens = vivi_alloc(cap * sizeof(size_t));
	if (!ids || !strands || !lens) {
		vivi_dealloc(ids);
		vivi_dealloc(strands);
		vivi_dealloc(lens);
		return false;
	}
	if (p->count) {
		memcpy(ids, p->ids, p->count * sizeof(int));
		memcpy(strands, p->strands, p->count * sizeof(uint8_t *));
		memcpy(lens, p->lens, p->count * sizeof(size_t));
	}
	vivi_dealloc(p->ids);
	vivi_dealloc(p->strands);
	vivi_dealloc(p->lens);
	p->ids = ids;
	p->strands = strands;
	p->lens = lens;
	p->cap = cap;
	return true;
}

bool vivi_pool_add_gene(vivi_pool *p, int id, const uint8_t *strand, size_t slen,
	const char **err)
{
	if (!err) err = &pool_err_sink;
	if (!p) { *err = "expected pool"; return false; }
	if (!strand) { *err = "expected string"; return false; }
	if (!pool_reserve(p, p->count + 1)) { *err = "out of memory"; return false; }
	uint8_t *copy = vivi_alloc(slen ? slen : 1);
	if (!copy) { *err = "out of memory"; return false; }
	memcpy(copy, strand, slen);
	p->ids[p->count] = id;
	p->strands[p->count] = copy;
	p->lens[p->count] = slen;
	p->count++;
	return true;
}

bool vivi_pool_add_chromosome(vivi_pool *p, const uint8_t *strand, size_t slen,
	const char **err)
{
	if (!err) err = &pool_err_sink;
	genome_scan_result res;
	if (!genome_gene_scan(&res, strand, slen, err)) return false;
	bool ok = true;
	for (size_t i = 0; i < res.count && ok; i++) {
		const genome_gene *g = &res.genes[i];
		if (!g->crc_ok) continue;   /* only intact molecules enter the pool */
		ok = vivi_pool_add_gene(p, g->id, strand + g->offset, g->size, err);
	}
	genome_scan_free(&res);
	return ok;
}

bool vivi_pool_amplify(vivi_amp_result *out, const vivi_pool *pool, int id,
	const vivi_amp_opts *opts, const char **err)
{
	if (!err) err = &pool_err_sink;
	memset(out, 0, sizeof(*out));
	out->id = -1;
	if (!pool || pool->count == 0) { *err = "empty pool"; return false; }
	vivi_amp_opts def = { 0.0, 0.0, 0, { 0.0, 0.0, 0.0, 0.0, 0 } };
	if (!opts) opts = &def;
	if (!(opts->p_access >= 0.0 && opts->p_access <= 1.0)
		|| !(opts->p_cross >= 0.0 && opts->p_cross <= 1.0)) {
		*err = "invalid probability";
		return false;
	}
	size_t target = SIZE_MAX;
	for (size_t i = 0; i < pool->count; i++)
		if (pool->ids[i] == id) { target = i; break; }
	if (target == SIZE_MAX) { *err = "target not in pool"; return false; }
	vivi_prng rng;
	vivi_prng_init(&rng, opts->seed);
	if (vivi_prng_chance(&rng, opts->p_access)) {
		out->read.dropped = 1;
		return true;
	}
	size_t pick = target;
	if (pool->count > 1 && vivi_prng_chance(&rng, opts->p_cross)) {
		pick = (size_t)(vivi_prng_next(&rng) % pool->count);
		if (pool->ids[pick] == id) pick = (pick + 1) % pool->count;
	}
	vivi_channel_opts ch = opts->ch;
	ch.seed = opts->seed + (uint32_t)pick * 0x9E37'79B9u + 1u;
	if (!vivi_channel_read(&out->read, pool->strands[pick], pool->lens[pick], &ch, err))
		return false;
	out->id = pool->ids[pick];
	return true;
}

File: framework/luau/pool.lua
--!native
local channel = require("./channel")
local genome = require("./genome")
local sub = string.sub

local TWO32 = 4294967296
local CROSS_MUL = 0x9E3779B9

local function new()
	return { ids = {}, strands = {} }
end

local function add_gene(p, id, strand)
	p.ids[#p.ids + 1] = id
	p.strands[#p.strands + 1] = strand
	return true
end

local function add_chromosome(p, strand)
	local genes, err = genome.scan(strand)
	if not genes then return nil, err end
	for i = 1, #genes do
		local g = genes[i]
		if g.crc_ok then
			add_gene(p, g.id, sub(strand, g.offset, g.offset + g.size - 1))
		end
	end
	return true
end

local function amplify(p, id, opts)
	opts = opts or {}
	local p_access = opts.p_access or 0.0
	local p_cross = opts.p_cross or 0.0
	local seed = opts.seed or 0
	local ch = opts.ch or {}
	local p_primer = opts.p_primer or 0.0
	local coverage = opts.coverage or 1
	local n = #p.ids
	if n == 0 then return nil, "empty pool" end
	if not (p_access >= 0.0 and p_access <= 1.0)
		or not (p_cross >= 0.0 and p_cross <= 1.0)
		or not (p_primer >= 0.0 and p_primer <= 1.0) then
		return nil, "invalid probability"
	end
	local target
	for i = 1, n do
		if p.ids[i] == id then
			target = i
			break
		end
	end
	if not target then return nil, "target not in pool" end
	local rng = channel.prng_new(seed)
	if channel.prng_chance(rng, p_access) then
		return { read = { strand = "", bases = 0, dropped = 1 }, id = -1 }
	end
	if p_primer > 0.0 and channel.prng_chance(rng, p_primer) then
		return { read = { strand = "", bases = 0, dropped = 1 }, id = -1 }
	end
	local pick = target
	if n > 1 and channel.prng_chance(rng, p_cross) then
		pick = channel.prng_next(rng) % n + 1
		if p.ids[pick] == id then pick = pick % n + 1 end
	end
	local c = {}
	for k, v in pairs(ch) do c[k] = v end
	c.seed = (seed + (pick - 1) * CROSS_MUL % TWO32 + 1) % TWO32
	local rd, err = channel.consensus_read(p.strands[pick], { ch = c, coverage = coverage })
	if not rd then return nil, err end
	return { read = rd, id = p.ids[pick] }
end

return {
	new = new,
	add_gene = add_gene,
	add_chromosome = add_chromosome,
	amplify = amplify,
}
